Scope: use direct iterator to checkNoAbstractMembers

diesalbla opened the pull request:
When profiling a compilation, I noticed that the `checkNoAbstractMembers` was the source of 6% of all the `::` allocations. 

Now, Scope implements a sort of single-linked list of `ScopeEntry` nodes. Now, its default iterator needs to traverse that list in reversed order, for which it needs to keep a list (stack) to LIFO those elements. Creating that stack is the source of those `::` allocations. 

It turns out that, as long as we only want to do a forall or exists with the iterator, as in `checkNoAbstractMethods`, the order does not matter. Thus: 

- We introduce a simple iterator for the `Scope` class that, unlike the
normal iterator, processes the elements as reachable through the
entries, and thus needs no extra list (stack) to LIFO its elements.
- We use this iterator in the `checkNoAbstractMethods`.

diesalbla committed:
Scope: use direct iterator to checkNoAbstractMembers

We introduce a simple iterator for the Scope class that, unlike the
normal iterator, processes the elements as reachable through the
entries, and thus needs no extra list (stack) to LIFO its elements.
As long as what we want to do with the iterator is a forall or exists,
as is the case in `checkNoAbstractMethods`, the order matters not.

The motivation was that we noticed `checkNoAbstractMembers` to be a
source of allocations of `::` objects, due to use of `Scope.toList`
from `Scope.partition`.

joshlemer commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -594,7 +594,7 @@ abstract class RefChecks extends Transform {
           // Avoid spurious duplicates: first gather any missing members.
           def memberList = clazz.info.nonPrivateMembersAdmitting(VBRIDGE)
-          val (missing, rest) = memberList partition (m => m.isDeferred && !ignoreDeferred(m))
+          val missing = memberList.reverseIterator.filter(m => m.isDeferred && !ignoreDeferred(m)).toList.reverse
```
Maybe we can avoid even more work with something like...

```scala
          val (missing, rest) = {
            val missing = ArrayBuffer[global.Symbol]()
            val rest = ArrayBuffer[global.Symbol]()

            memberList.reverseIterator.foreach { s =>
              if (s.isDeferred && !ignoreDeferred(s)) {
                missing.addOne(s)
              } else if (s.isAbstractOverride && s.isIncompleteIn(clazz)) {
                rest.addOne(s)
              }
            }

            (missing, rest)
          }
```

...

```scala
rest.reverseIterator.foreach { member => 
  ...
```
...
```scala
          for (member <- missing.reverseIterator ; msg = diagnose(member) ; if msg != null) {
...
```

diesalbla replied:
So, I did not thought at first about optimising this part because these are lists of errors, which are always empty for correct compilations. 

What I have done at the end is to use mutable `List` variables, and use the `::=` operator for the LIFO behaviour. This has the benefit of having no allocations (`Nil`) in the common, no-errors case.

diesalbla replied:
It may be possible to also add the `grouped` and `missingMethods` logic, for consolidating pairs of setter-getter into a single element, but I am not sure it would be worth as much.

retronym commented on src/reflect/scala/reflect/internal/Scopes.scala:
```diff
@@ -418,6 +431,8 @@ trait Scopes extends api.Scopes { self: SymbolTable =>
      */
     def iterator: Iterator[Symbol] = toList.iterator
 
+    def reverseIterator: Iterator[Symbol] = new ScopeIterator(this)
```
We should document that this when using `reverseIterator` is not safe to mutate the `Scope`. Some uses of `Scope#iterator` do mutate the scope during iteration, that's another reason that the list is created.

diesalbla committed:
Attend code review comments.

- In Scope we add some documentation and warning about the
  new operation.
- In `RefChecks`, We generate the `missing` and `rest` lists
  directly from a single pass of the iterator. Using the `::=`
  operator we get the "LIFO" behaviour we need to reverse.

retronym reviewed (approved):
I like it!

It's worth using `reverseIterator` will be a win when dealing with a temporary `Scope` (like the one returned by `findMembers` et al, so long as that's the only operation used and `toList` is never called.

`Scope`-s are also used for the class declarations and for block scopes. In those cases, it might be harder to avoid the `toList` call altogether.
